### PR TITLE
Calendar

### PR DESCRIPTION
**Pull request by ThePotatoGuy**
#37 #1538 

The calendar is _finally_ here.

morewhiteresult

# Main features:
* Visible representation of events / other things that correspond to calendar days.
* User-friendly way to select a date.

# Additional Features:
* Calendar is clickable from the idle mode.
* Users can now set the date they started dating Monika.
* If a calendar day shows more than 3 events, the day becomes clickable and a screen with a list of all the events for that day is shown when clicked.

## Dialogue flow:
There are 2 main additions here:

### First time clicking calendar:
1. Player clicks calendar for the first time
2. Monika comments on how cool the calendar is and shows player.
3. Flag is set that enables the "When did we start dating" event

### When did we start dating:
This event is a pool question that has Monika inform the player of when they started dating. If the player has never seen this event before, Monika will allow the player to pick the dating start date. **This can only be done once. If you mess this up, we will not fix it for you.**
This event unlocks itself 30 days after `first_session`.

1. The diff between today and `first_session` is calculated.
    1. If the diff is 0 (which means today is the first session), monika remarks how this is impossible and the event ends. The event is locked for 30 days.
    2. otherwise, continue
2. If this is the first time this event has happened, go to step 4.
3. Otherwise, Monika tells player when they started dating. The event ends.
4. Monika mentions the currently stored value for `first_session`. She implies that her memory is probably wrong.
5. Monika asks player if this date is correct.
    1. Player says yes. Monika is glad. go to step 12.
    2. Player says No. Monika apologizes and asks player when yall started dating.
6. Calendar selection screen appears.
7. User selects date (or not).
    1. If user does not select date or picks the same date as the one currently stored. go to step 8.
    2. If user selects today's date, goto step 9.
    3. If user selects a future date, goto step 10.
    4. Otherwise go to step 11.
8. Monika is confused why player would pick same date / not pick a date.
    1. Monika asks player if `first_session` is the wrong date.
        1. If user says its wrong date, Monika tells user to pick correct date. Jump back to step 6.
            * Upon the third time this happens, Monika suggests they do this another time. Event is locked for 30 days. Event ends.
            * **NOTE: the counter for this is shared with step 9.1**
        2. If user says it's correct date, goto step 11.
9. Monika is confused why player would pick today.
    1. Monika scolds player for not taking this seriously.
        * Upon the third time this happens, Monika suggests they do this another time. Event is locked for 30 days. Event ends.
        * **NOTE: the counter for this is shared with step 8.1.1**
    2. Jump back to step 6
10. Monika is confused why player picks a future date.
    * Upon the 2nd time this happens, Monika suggests they do this another time. Event is locked for 30 days. Event ends.
    1. Monika asks if they haven't been dating this whole time.
        1. If user says it was a mislick, Monika is relieved. Jump back to step 6
        2. If user says that you werent dating, Monika is further confused and closes her eyes.
            1. If user says they are kidding, Monika says that's a mean joke. Asks player not to do that again. Jump back to step 6.
            2. If user does not say anything or click any choice, Monika believes it to be true. A flag is set, marking that yall are friends and nothing more. This will be leveraged in Affeection. This event is locked for the time being. Event ends. **NOTE: Even if this flow happens, other topics are UNCHANGED. We will handle all of that shit in affection.**
11. Monika confirms with player that selected date is correct.
    1. User confirms. Monika asks one last time, implying that you cannot change this again.
        1. User confirms. The selected date is returned. go to step 12.
        2. User does not confirm. Monika knew you were unsure. Jump back to step 6.
            * Upon the third time this happens, Monika suggests they do this another time. Event is locked for 1 day. Event ends.
            * **NOTE: the counter for this is shared with 11.2**
    2. User does not confirm. Monika asks player to reselect. Jump back to step 6.
        * Upon the third time this happens, Monika suggests they do this another time. Event is locked for 1 day. Event ends.
        * **NOTE: the counter for this is shared with 11.1.2**
12. Anniversaries are recalculated and unlocked accordingly. The calendar database is readjusted. The selected date becomes saved as the new `first_session`.
13. The persistent is saved. A flag is set saying that we have confirmed the start date.
14. Event ends.

# API:
There are 2 main modes for the calendar API:
1. View mode, which shows the calendar and events that are on each day.
2. Selection mode, which allows you to pick a day from the calendar. (This returns a `datetime`)

## Calendar items:
The things that appear on the calendar are known as `Calendar Items`.
They come in 2 flavors (for now):
1. An `Event` object.
2. A `Repeatable` type.

The `Repeatable` type is a new type of event that does not correspond to an `eventlabel` nor an `Event` object. It's purely for marking something on the calendar.

## Storage:
Calendar items are stored in a layered-dict structure, known as the `calendar database`:
```
month: {
    day: {
        key: {
            tuple:
                [0]: type of calendar item
                [1...]: additional data items related to the calendar item type.
        }
    }
}
```
1. The outermost layer is a dict with **month** as the keys. (`range(1,13)`)
2. The next layer (the value each **month** corresponds to) is a dict with **day** as the keys. (`range(1,32)`)
3. The next layer (the value each **day** corresponds to) is a dict with a special identifier as the keys.
    * If the calendar item is an `Event`, the identifier is the `eventlabel`.
    * If the calendar item is a `Repeatable`, the identifier is chosen by the caller. **You should aim for this to be unique**
4. Each item in this dict is a tuple whose values vary depending on the type of the calendar item.
    * the first element in this tuple is the calendar item type. `CAL_TYPE_EV` for event and `CAL_TYPE_REP` for Repeatable.
    1. For an `Event` type, the only item in the tuple is the first item, which is `CAL_TYPE_EV`.
        1. `ev.label` is used as display text on the calendar.
        2. `ev.years` is used for the year repeating system.
    2. For a `Repeatable` type:
        1. `[0]` - `CAL_TYPE_REP`
        2. `[1]` - label to display on the calendar
        3. `[2]` - list of years to use for the year repeating system

### Year repeating
The calendar database represents a single year of events. Events that need to repeat multiple years use a list of years that event repeats on. For `Event`, that list is it's `years` property. For a `Repeatable`, that list is given by you when you add it to the calendar.

If the year list is None, it is assumed the event does **not** repeat. (If the item is an `Event`, the `start_date` is used for figuring out the correct year to display the event.)

If the year list is an empty list, it is assumed the event repeats **annually**.

Otherwise, the year list is assumed to be a list of numbers, which are the years.

### Other API:
The `mas_calendar` store contains a variety of helper functions for adding / removing events from the calendar database. It is **HIGHLY RECOMMENDED** to use those functions instead of manually adding stuff to the database.
Here is a list of all the functions:
**Refer to zz_calendar.rpy for detailed documentation**
```
genFriendlyDisplayDate - generates a display-friendly string representation of a given datetime
addEvent - adds an Event to the calendar, using the Event's properties to decide where to add it.
addEvent_evd - adds an Event to the calendar, using the given date object
addEvent_evdt - adds an Event to the calendar, using the given datetime object
addRepeatable - adds a Repeatable to the calendar.
addRepeatable_d - adds a Repeatable to the calendar, using a given date object
addRepeatable_dt - adds a Repeatable to the calendar, using a given datetime object
removeEvent - removes an Event from teh calendar, using the Event's properties to find out where to remove it.
removeEvent_eld / removeEvent_evd - removes an Event from teh calendar, using a given date object to find out where to remove it.
removeEvent_eldt / removeEvent_evdt - removes an Event from the calendar, using a given datetime object to find out where to remove it.
removeEvent_el / removeEvent_ev - removes an Event from the calendar, capable of doing month / day searches as well as full searches for removal.
removeRepeatable - removes a Repeatable from the calendar, capable of doing month / day searches as well as full searches for removal.
removeRepeatable_d - removes a Repeatable from the calendar, using a given date object to find out were to remove it.
removeRepeatable_dt - removes a Repeatble from the calendar, using a given datetime object to find out where to remove it.
```

We now have a singular database for accessing **all** events. This will make processing all events easier and also facilitates `Event` lookup in the calendar.
Here is a list of functions regarding this singular database:
**Refer to event-handler for detailed documentation**
```
mas_getEV - returns an Event given the label
mas_getEVCL - returns an Event's display label given the label
```
Having this singular database means that all `Event` objects, not just the `evhand` ones, can get their `shown_count` and `last_seen` properties tracked in `call_next_event`

Anniversaries now have their own database, but are still stored in the `evhand` eventdatabase as well. 
The new `mas_anni` store contains this database as well as a few functions for resetting / cleaning the anniversaries. 
**Refer to script-anniversary for more information**

The month/year/start_of_day adjustment functions have been moved from script-anniversaries to the `mas_utils` store.

The `isPast`, `isFuture`, `isPresent` (the relative date functions) have been internalized in `evhand` so they can be accessed in stores. The global functions serve as pointers to the internalized ones.

# Calendar data
All calendar data is saved in an external json located in the same directory as `persistent`. Separating the two will help with preventing the save data loss bug from affecting a calendar full of events.

### CODE REVIEW

File: Monika After Story/game/definitions.rpy [hunk 1]
```diff
@@ -662,41 +671,72 @@ python early:
...
+                        continue
 
 
-                if event_time and events[ev].action is not None:
+                if e.action is not None:
                     #Perform the event's action
-                    if events[ev].action == EV_ACT_PUSH:
+                    if e.action == EV_ACT_PUSH:
                         pushEvent(ev)
                     elif events[ev].action == EV_ACT_QUEUE:
```
Comment (ThePotatoGuy) [inline Monika After Story/game/definitions.rpy]: @aldoram5 there's an inconsistency here, why is this line not using `e.action`

File: Monika After Story/game/script-topics.rpy [hunk 2]
```diff
@@ -6619,3 +6619,289 @@ label monika_hair_down:
...
+                        hide screen mas_background_timed_jump
+                        # wow what a mean joke
+
+                        if no_dating_joke:
+                            # you only get this once per thru
+                            jump .had_enough
+
+                        # otherwise mention that this was mean
+                        m 2tfc "[player]!"
+                        m 2rksdlc "That joke was little mean."
```
Comment (juliathegolden) [inline Monika After Story/game/script-topics.rpy]: ` +                        m 2rksdlc "That joke was little mean." `
Try "That joke was a little mean."

File: Monika After Story/game/zz_calendar.rpy [hunk 3]
```diff
@@ -0,0 +1,1693 @@
...
+                            background None
+                            hover_sound gui.hover_sound
+                            activate_sound gui.activate_sound
+
+                            action Return(final_item[1])
+
+
+label _first_time_calendar_use:
+    m 1eub "Oh, I see you noticed that pretty calendar hanging in the wall, [player]"
+    m "It helps me to keep track of important events, ehehe~"
```
Comment (juliathegolden) [inline Monika After Story/game/zz_calendar.rpy]: +    m 1eub "Oh, I see you noticed that pretty calendar hanging in the wall, [player]"
+    m "It helps me to keep track of important events, ehehe~" ``

I'd change 'hanging in' to "hanging on" and change "helps me to keep" to "helps me keep".

Comment (juliathegolden) [inline Monika After Story/game/zz_calendar.rpy]: +    m 1eub "Oh, I see you noticed that pretty calendar hanging in the wall, [player]"
+    m "It helps me to keep track of important events, ehehe~" ``

I'd change 'hanging in' to "hanging on" and change "helps me to keep" to "helps me keep".

File: Monika After Story/game/script-topics.rpy (hunk 2 above)
Comment (juliathegolden) [inline Monika After Story/game/script-topics.rpy]: + m 2rksdlc "That joke was little mean."
Try "That joke was a little mean."

File: Monika After Story/game/script-topics.rpy [hunk 4]
```diff
@@ -6619,3 +6619,289 @@ label monika_hair_down:
...
+                        m 2eksdlc "You really had me worried there."
+                        m "Don't play around like that okay?"
```
Comment (juliathegolden) [inline Monika After Story/game/script-topics.rpy]: +                        m "Don't play around like that okay?" ``
comma after that

File: Monika After Story/game/definitions.rpy (hunk 1 above)
Comment (aldoram5) [reply Monika After Story/game/definitions.rpy]: fixed, I need to get a new pair of glasses ~~That's true though broke them yesterday~~

File: Monika After Story/game/zz_calendar.rpy [hunk 5]
```diff
@@ -0,0 +1,1695 @@
...
+        CALENDAR_MONTH_INCREASE = "MONTH_INCR" # signals to increase the current selected month
+        CALENDAR_MONTH_DECREASE = "MONTH_DECR" # signals to decrease the current selected month
+        CALENDAR_YEAR_INCREASE = "YEAR_INCR" # signals to increase the current selected year
+        CALENDAR_YEAR_DECREASE = "YEAR_DECR" # signals to decrease the current selected month
+
+        # Color used for the day number
+        TEXT_DAY_COLOR = "#000000" # PINK: "#ffb0ed"
+
+        # Month names constant array
+        MONTH_NAMES = ["Unknown", "January", "Febuary",
```
Comment (juliathegolden) [inline Monika After Story/game/zz_calendar.rpy]: `+ MONTH_NAMES = ["Unknown", "January", "Febuary",`
Should be February

File: Monika After Story/game/zz_calendar.rpy [hunk 6]
```diff
@@ -0,0 +1,1695 @@
...
+        """
+        removeRepeatable_d(identifier, _datetime.date())
+
+
+# add repeatable events
+init python:
+
+    import store.mas_calendar as calendar
+
+    calendar.addRepeatable("New years day","New years day",month=1,day=1,year_param=list())
```
Comment (juliathegolden) [inline Monika After Story/game/zz_calendar.rpy]: All of the words in the Holidays should be capitalized, so "New Year's Day, Valentine's Day, White Day, Christmas Eve, Christmas Day, New Year's Eve.

**Comment by jmwall24** [comment]
There is one small thing I've been thinking about in regards to changing the date you started dating. 
 Depending on when the calendar comes out, if it's after say your 6 month anniversary based on the actual date you started dating her, but before the 6 month anniversary of the current wrong date you'd never get to see the 6 month anniversary dialogue.  Not a huge deal but still kinda sucks to miss it altogether.  So if this were the case, could you keep telling her you're not sure everyday until you got to the anniversary and then set it to the correct date?  Is there a limit to how many times you can tell her you're not sure before it locks it forever?

**Comment by ThePotatoGuy** [comment]
@jmwall24 all the anniversaries will adjust themselves and unlock themselves when you change the date. You will be able to view the past anniversaries in `Repeat Conversation`

**Comment by jmwall24** [comment]
Ah ok, didn't know they were available in repeat convos, thanks.

**Pull request merged**